debug_scripts/estimate_epoch_start_time.py

```diff
@@ -0,0 +1,142 @@
+import requests
+import time
+import math
+import argparse
+
+
+# Function to get block data by block hash
+def get_block_by_hash(url, block_hash):
+    payload = {
+        "jsonrpc": "2.0",
+        "id": "dontcare",
+        "method": "block",
+        "params": {
+            "block_id": block_hash
+        }
+    }
+    response = requests.post(url, json=payload)
+    return response.json()['result']['header']
+
+
+# Function to fetch epoch lengths for the past n epochs and calculate the weighted average using exponential decay
+def get_exponential_weighted_epoch_lengths(url,
+                                           starting_epoch_hash,
+                                           num_epochs,
+                                           decay_rate=0.1):
+    epoch_lengths = []
+    current_hash = starting_epoch_hash
+
+    for i in range(num_epochs):
+        # Get the block data by hash
+        block_data = get_block_by_hash(url, current_hash)
+
+        # Get the timestamp of this block (start of current epoch)
+        current_timestamp = int(block_data['timestamp'])
+
+        # Get the next epoch hash (start block hash of current epoch)
+        next_epoch_hash = block_data['next_epoch_id']
+
+        # Fetch the block data for start of previous epoch
+        next_block_data = get_block_by_hash(url, next_epoch_hash)
+        next_timestamp = int(next_block_data['timestamp'])
+
+        # Calculate the length of the epoch in nanoseconds
+        epoch_length = current_timestamp - next_timestamp
+        epoch_length_seconds = epoch_length / 1e9  # Convert to seconds
+        epoch_lengths.append(epoch_length_seconds)
+
+        print(f"Epoch {i+1}: {epoch_length_seconds} seconds")
+
+        # Move to the next epoch
+        current_hash = next_epoch_hash
+
+    # Apply exponential decay weights: weight = e^(-lambda * i), where i is the epoch index and lambda is the decay rate
+    weighted_sum = 0
+    total_weight = 0
+    for i in range(num_epochs):
+        weight = math.exp(-decay_rate * i)
+        weighted_sum += epoch_lengths[i] * weight
+        total_weight += weight
+
+    # Calculate the weighted average using exponential decay
+    exponential_weighted_average_epoch_length = weighted_sum / total_weight
+
+    print(
+        f"\nExponential weighted average epoch length: {exponential_weighted_average_epoch_length} seconds"
+    )
+
+    return epoch_lengths, exponential_weighted_average_epoch_length
+
+
+# Function to approximate future epoch start dates
+def predict_future_epochs(starting_epoch_timestamp, avg_epoch_length,
```

Comment:
One observation - the tool predicts that the next epoch will start in 11 hours, but debug-ui says it'll happen in 13 hours.
I wonder where the discrepancy comes from :thinking: Maybe debug-ui doesn't use exponential average. /cc @robin-near